### ISSUE TITLE
Using SIMD for dealing with json at speed

### ISSUE THREAD

**Issue by hpvd** [apache/pinot#13773]
Using SIMD for dealing with json at speed

inspired by postgreSQL:
see: https://www.phoronix.com/news/PostgreSQL-Opt-JSON-Esc-SIMD

and since more and more CPUs support AVX512 or its successors:
https://www.phoronix.com/review/simdjson-avx-512
https://simdjson.org/

**Comment by abhioncbr** [comment]
@hpvd, what would you suggest, using the simdjson library for all JSON data handling or something else?

**Comment by hpvd** [comment]
I think this would be a multistep approach. We can look what is possible on https://simdjson.org/ and just pick one place in Pinot and give it a try. In the end we can utilize it in many ways..

**Comment by siddharthteotia** [comment]
@hpvd  @abhioncbr - I have been very interested in exploring more wide and holistic use of SIMD in Pinot. Historically, that endeavor has  not been successful because of no support in Java for the low level primitives. JNI is of course an option.

For this issue, how are you planning to use SIMD in Pinot code base ? Is it via the JNI bridge that we build over Intel compiler intrinsics or using an abstraction (e.g JDK vector APi available in 14 onwards IIRC) or something else ?

**Comment by siddharthteotia** [comment]
My high level suggestion would be that if there is indeed a possible path to leverage SIMD acceleration in JAVA, rather than doing piece-wise work for a specific scenario, it would be better to first get a handle on how it will be integrated into Pinot code base so that we can also re-use them in more appropriate places (e.g in the query engine). Also need to evaluate the portability aspect as well. 

> We can look what is possible on https://simdjson.org/ and just pick one place in Pinot and give it a try. In the end we can utilize it in many ways..

Agree with POCing one aspect but when we actually decide to build the feature, it should ideally be done thinking of broader, long term use thinking about everything since we are likely going to introduce platform specific dependencies into the codebase.

**Comment by hpvd** [comment]
@siddharthteotia have you already looked into this one:
https://github.com/simdjson/simdjson-java

**Comment by hpvd** [comment]
We may also look into Apache Doris how they leverage it....

**Comment by abhioncbr** [comment]
Yes, my understanding was also to use the simd Java bindings. As @hpvd suggested, we can explore how jdk based projects are using it and we can take a path forward based on that.

**Comment by siddharthteotia** [comment]
> we can explore how jdk based projects are using it and we can take a path forward based on that.

+1. Yes let's do some survey

> https://github.com/simdjson/simdjson-java

This is based on incubator version of vector support in JDK (Project Panama by Open JDK AFAIK). Note that the package still says "incubator" so I am not sure of production use / support for this. We have done this in the past where we took a dependency on less than productionized library (Lbuffer) and it proved to be unstable once in a while. Recently we have removed it. 

So, I think as a first step it will be good to see if any of the latest versions of JDK actually support it or not before we go way deeper in the POC / performance evaluation with above library

Take a look at project Gandiva (under Arrow) too. We can also build a JNI bridge ourselves. 

I think the investment really depends on some value via POC.

Curious if @gortiz  / @richardstartin  have any advice / suggestions.

**Comment by hpvd** [comment]
this reads is already one year old, but pretty interesting: it shows how elastic / lucene leverage SIMD, handle incubating possibilities, show some benchmarks etc.
https://www.elastic.co/de/blog/accelerating-vector-search-simd-instructions

**Comment by hpvd** [comment]
this includes history, state and goals of vector API in java:
https://openjdk.org/jeps/469

**Comment by kishoreg** [comment]
This is a fantastic initiative and +100 on getting native SIMD. Given the pace at which Java is moving, it might be a good idea to slowly extract interfaces where SIMD can benefit. This will allow users/companies to stay with older jdk while other companies can move forward.

we don't want to stuck in the same mode as last time where moving out of Java 8 meant waiting for all users to migrate to Java 8.

**Comment by hpvd** [comment]
jep would be great, if we find a way were the people who want and can (internal restrictions, used hardware selection..) are able to benefit from new possibilities without having to wait till everybody is ready.

**Comment by hpvd** [comment]
just edited the title to `SIMD for dealing with json *(and more)* at speed` :-)

**Comment by gortiz** [comment]
> Curious if @gortiz / @richardstartin have any advice / suggestions.

I think explorations in this area are very interesting, but AFAIK Panama is not fast enough yet. Last month in JCrete we were discussing about how to access native code efficiently and it looks like nothing has changed (yet). Calling JNI/Panama code per row is prohibitively slow. The good news is that in single-stage engine and in the leaf stages in multi-stage engine these calls can be done at block level, so we should be able to absorb the cost of the JNI call.

**Comment by hpvd** [comment]
good overview and starter:
SIMD Parallel Programming with the Vector API By José Paumard

> This session explains the differences between parallel streams and parallel computing, and how SIMD computations are working internaly on simple examples. It then shows the patterns of code that the Vector API is giving along with their performances, and how you can use them to improve your in-memory data processing computations. More advanced techniques are also presented, to go beyond the basic examples.

https://www.youtube.com/watch?v=36DN9sE7ja4

includes usecases and basic speed comparisons:
![2024-08-12_11h42_49](https://github.com/user-attachments/assets/a485cda8-c31e-4956-b1a4-66d143e06be5)
![2024-08-12_11h42_24](https://github.com/user-attachments/assets/c6f8caf6-1a10-42d5-97d5-7cf5d799e749)

**Comment by hpvd** [comment]
just to get an understanding how other projects handle this:
for apache lucene, using more SIMD in an easy way is one of the reasons to make java v21 mandatory with the upcoming next major release  of lucene (v10)

V

> ectorization
> 
> Parallelism and concurrency, while distinct, often translate to "splitting a task so that it can be performed more quickly", or "doing more tasks at once". Lucene is continually looking at new algorithms and striving to implement existing ones in more performant and efficient ways. One area that is now more straightforward to us in Java is data level parallelism - the use of SIMD (Single Instruction Multiple Data) vector instructions to boost performance.
> 
> Lucene is using the latest JDK Vector API to implement vector distance computations that result in efficient hardware specific SIMD instructions. These instructions, when run on supporting hardware, can perform floating point dot product computations 8 times faster than the equivalent scalar code. This [blog](https://www.elastic.co/blog/accelerating-vector-search-simd-instructions) contains more specific information on this particular optimization.
> 
> With the move to Java 21 minimum, it is a lot more straightforward to see how we can use the JDK Vector API in more places. We're even experimenting with the possibility of calling customized SIMD implementations with FFI, since the overhead of the native call is now quite minimal.

https://www.elastic.co/search-labs/blog/lucene-and-java-moving-forward-together

**Comment by hpvd** [comment]
as expected, lucene changes requirements and v10 now requires Java 21, see https://lucene.apache.org/core/corenews.html#apache-lucenetm-1000-available

**Comment by hpvd** [comment]
just started a list to get an overview of things we are missing with staying using/being compatible to older Java versions,
and determine the right point of time when its maybe worth to drop one or find a way to work around it:
https://github.com/apache/pinot/issues/14325

**Comment by chenboat** [comment]
From our experience at Uber, when we used Java 21 Runtime, many Pinot functions are auto-vectorized by the JIT Hotspot compiler. These include both Pinot core functions and also Lucene related lib.

Core Pinot code
<img width="1238" height="313" alt="Image" src="https://github.com/user-attachments/assets/0c3c653a-a661-4b93-9295-2175c8efcd73" />

Lucene code

<img width="1398" height="466" alt="Image" src="https://github.com/user-attachments/assets/a5b30c59-82a6-49ef-9efb-584b8d554f0a" />
